[WebVTT] Modernize VTTCue

jernoble opened the pull request:
#### 687a6e6a1e60b9f64c6892feee42156e959508d8
<pre>
[WebVTT] Modernize VTTCue
<a href="https://bugs.webkit.org/show_bug.cgi?id=258578">https://bugs.webkit.org/show_bug.cgi?id=258578</a>
rdar://111393504

Reviewed by Eric Carlson.

VTTCue has a number of stylistic and functional implementation details that
are held over from its original implementation in 2014, and could benefit from
being brought up to a more modern set of practices and language features used
in other parts of WebKit.

The bindings generator now supports IDL-defined enumerations, and automatically
converts between them and C++ defined enumerations, so there is no need to
have string checks exist in VTTCue for properties that take those enumerations.

C++ enumerations should have the same type name as IDL enumerations, and values
for those enumerations should be equivalent as well (after converting to
CamelCase from snake-case). Those enumerations should therefore be `enum class`.

The existing enumerations included sentinel values for static checks of the
enumerations&apos; sizes. Add support in EnumTraits for calculating the minimum,
maximum, and count of values at compile time, making sentinal values unnecessary.

The WebVTT specification now defines heights and font sizes in terms of &apos;vh&apos; and
&apos;vw&apos;, which are percentages of the video viewport height and width, respectively.
Now that WebKit supports these CSS units (as well as &apos;cqh&apos; and &apos;cqw&apos;, which are
the equivalent units for percentage of container height and width), it is no
longer necessary to pass the layout size of the video element into the VTTCue
(and its subclasses) to calculate the CSS values of cue properties.

The WebVTT specification added more default CSS properties which need to be set
on the resulting elements, so those are added here too.

Add support for lineAlign and positionAlign properties in layout, where those
properties were already parsed and exposed though DOM APIs.

* Source/WTF/wtf/EnumTraits.h:
(WTF::EnumValues::std::max):
(WTF::EnumValues::std::min):
(WTF::EnumValues::forEach):
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateDisplay):
(WebCore::MediaControlTextTrackContainerElement::processActiveVTTCue):
(WebCore::MediaControlTextTrackContainerElement::updateActiveCuesFontSize):
* Source/WebCore/html/track/InbandGenericTextTrack.cpp:
(WebCore::InbandGenericTextTrack::updateCueFromCueData):
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::TextTrackCue::getDisplayTree):
(WebCore::TextTrackCue::setFontSize):
* Source/WebCore/html/track/TextTrackCue.h:
(WebCore::TextTrackCueBox::applyCSSProperties):
* Source/WebCore/html/track/TextTrackCueGeneric.cpp:
(WebCore::TextTrackCueGenericBoxElement::applyCSSProperties):
(WebCore::TextTrackCueGeneric::setFontSize):
(WebCore::TextTrackCueGeneric::setLine): Deleted.
* Source/WebCore/html/track/TextTrackCueGeneric.h:
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCueBox::applyCSSProperties):
(WebCore::VTTCue::VTTCue):
(WebCore::VTTCue::setVertical):
(WebCore::VTTCue::setLine):
(WebCore::VTTCue::setLineAlign):
(WebCore::VTTCue::position const):
(WebCore::VTTCue::setPositionAlign):
(WebCore::VTTCue::setAlign):
(WebCore::VTTCue::calculateComputedLinePosition const):
(WebCore::VTTCue::calculateComputedTextPosition const):
(WebCore::VTTCue::calculateComputedPositionAlignment const):
(WebCore::VTTCue::calculateMaximumSize const):
(WebCore::VTTCue::calculateDisplayParameters):
(WebCore::VTTCue::obtainCSSBoxes):
(WebCore::VTTCue::updateDisplayTree):
(WebCore::VTTCue::getDisplayTree):
(WebCore::VTTCue::getPositionCoordinates const):
(WebCore::VTTCue::setCueSettings):
(WebCore::VTTCue::getCSSAlignment const):
(WebCore::VTTCue::getCSSWritingMode const):
(WebCore::VTTCue::setFontSize):
(WebCore::VTTCue::toJSON const):
(WebCore::horizontalKeyword): Deleted.
(WebCore::VTTCue::initialize): Deleted.
(WebCore::VTTCue::vertical const): Deleted.
(WebCore::VTTCue::lineAlign const): Deleted.
(WebCore::VTTCue::positionAlign const): Deleted.
(WebCore::VTTCue::align const): Deleted.
* Source/WebCore/html/track/VTTCue.h:
(WebCore::VTTCue::vertical const):
(WebCore::VTTCue::lineAlign const):
(WebCore::VTTCue::positionAlign const):
(WebCore::VTTCue::align const):
(WebCore::VTTCue::element const):
(WebCore::VTTCue::backdrop const):
(WebCore::VTTCue::fontSize const):
(WebCore::VTTCue::fontSizeIsImportant const):
(WebCore::VTTCue::left const):
(WebCore::VTTCue::top const):
(WebCore::VTTCue::width const):
(WebCore::VTTCue::height const):
(WebCore::VTTCue::getWritingDirection const): Deleted.
(WebCore::VTTCue::getAlignment const): Deleted.
* Source/WebCore/html/track/VTTCue.idl:
* Source/WebCore/rendering/RenderVTTCue.cpp:
(WebCore::RenderVTTCue::initializeLayoutParameters):
(WebCore::RenderVTTCue::placeBoxInDefaultPosition):
(WebCore::RenderVTTCue::overlappingObjectForRect const):
(WebCore::RenderVTTCue::shouldSwitchDirection const):
(WebCore::RenderVTTCue::moveBoxesByStep):
(WebCore::RenderVTTCue::findNonOverlappingPosition const):
(WebCore::RenderVTTCue::repositionCueSnapToLinesSet):
(WebCore::RenderVTTCue::repositionGenericCue):
(WebCore::RenderVTTCue::repositionCueSnapToLinesNotSet):
* Tools/TestWebKitAPI/Tests/WTF/EnumTraits.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/265596@main">https://commits.webkit.org/265596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c325006ff5c829238a0218017b578de353fe9dae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11325 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11877 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12976 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10798 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11346 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13915 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11520 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13708 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11487 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12366 "3 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13393 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9661 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10269 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17457 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/9621 "Built successfully and passed tests") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10730 "1 api test failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10425 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13636 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10746 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10840 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8921 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/11422 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10010 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3088 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2720 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14286 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11743 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10694 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2861 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -163,77 +160,94 @@ void VTTCueBox::applyCSSProperties(const IntSize& videoSize)
+    // NOTE: For 'top' and 'left', see step 25. in 7.2, Processing Cue Settings:
+    //   Let left be x-position vw and top be y-position vh.
+    // Use 'cqh' and 'cqw' rather than 'vh' and 'vw' here, as the video viewport
+    // is not a true viewport, but it is a container, so they serve right purpose.
```
Nit: “so they serve right purpose”?

jernoble replied:
Whoops. "serve the same purpose"

eric-carlson commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -163,77 +160,94 @@ void VTTCueBox::applyCSSProperties(const IntSize& videoSize)
+    //   If the WebVTT cue writing direction is horizontal, then let width be size
+    //   vw and height be auto. Otherwise, let width be auto and height be size vh.
+    // Use 'cqh' and 'cqw' rather than 'vh' and 'vw' here, as the video viewport
+    // is not a true viewport, but it is a container, so they serve right purpose.
```
Ditto

eric-carlson commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -798,127 +697,242 @@ double VTTCue::calculateComputedTextPosition() const
+        //    of the cue text is left-to-right, line-left otherwise.
+        return m_displayDirection == CSSValueLtr ? PositionAlignSetting::LineRight : PositionAlignSetting::LineLeft;
+    case AlignSetting::Center:
+        // 4. If the cue text alignment is center, return 50 and abort these steps.
```
Either this comment or the next line is wrong.

jernoble replied:
The comment is wrong; will fix.

eric-carlson commented on Source/WebCore/html/track/VTTCue.cpp:
```diff
@@ -798,127 +697,242 @@ double VTTCue::calculateComputedTextPosition() const
-    if (m_snapToLines && !m_displayPosition.first
-        && (m_writingDirection == VerticalGrowingLeft || m_writingDirection == VerticalGrowingRight))
-        m_displayPosition.first = 0;
+void VTTCue::obtainCSSBoxes()
```
“obtain” is odd, even if it does come from the spec text, maybe “update”?

jernoble replied:
I'd prefer to have the method have the same name as the spec text; it's going to make things easier for future implementors to figure out what's going on.

jernoble committed:
[WebVTT] Modernize VTTCue
https://bugs.webkit.org/show_bug.cgi?id=258578
rdar://111393504

Reviewed by Eric Carlson.

VTTCue has a number of stylistic and functional implementation details that
are held over from its original implementation in 2014, and could benefit from
being brought up to a more modern set of practices and language features used
in other parts of WebKit.

The bindings generator now supports IDL-defined enumerations, and automatically
converts between them and C++ defined enumerations, so there is no need to
have string checks exist in VTTCue for properties that take those enumerations.

C++ enumerations should have the same type name as IDL enumerations, and values
for those enumerations should be equivalent as well (after converting to
CamelCase from snake-case). Those enumerations should therefore be `enum class`.

The existing enumerations included sentinel values for static checks of the
enumerations' sizes. Add support in EnumTraits for calculating the minimum,
maximum, and count of values at compile time, making sentinal values unnecessary.

The WebVTT specification now defines heights and font sizes in terms of 'vh' and
'vw', which are percentages of the video viewport height and width, respectively.
Now that WebKit supports these CSS units (as well as 'cqh' and 'cqw', which are
the equivalent units for percentage of container height and width), it is no
longer necessary to pass the layout size of the video element into the VTTCue
(and its subclasses) to calculate the CSS values of cue properties.

The WebVTT specification added more default CSS properties which need to be set
on the resulting elements, so those are added here too.

Add support for lineAlign and positionAlign properties in layout, where those
properties were already parsed and exposed though DOM APIs.

* Source/WTF/wtf/EnumTraits.h:
(WTF::EnumValues::std::max):
(WTF::EnumValues::std::min):
(WTF::EnumValues::forEach):
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateDisplay):
(WebCore::MediaControlTextTrackContainerElement::processActiveVTTCue):
(WebCore::MediaControlTextTrackContainerElement::updateActiveCuesFontSize):
* Source/WebCore/html/track/InbandGenericTextTrack.cpp:
(WebCore::InbandGenericTextTrack::updateCueFromCueData):
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::TextTrackCue::getDisplayTree):
(WebCore::TextTrackCue::setFontSize):
* Source/WebCore/html/track/TextTrackCue.h:
(WebCore::TextTrackCueBox::applyCSSProperties):
* Source/WebCore/html/track/TextTrackCueGeneric.cpp:
(WebCore::TextTrackCueGenericBoxElement::applyCSSProperties):
(WebCore::TextTrackCueGeneric::setFontSize):
(WebCore::TextTrackCueGeneric::setLine): Deleted.
* Source/WebCore/html/track/TextTrackCueGeneric.h:
* Source/WebCore/html/track/VTTCue.cpp:
(WebCore::VTTCueBox::applyCSSProperties):
(WebCore::VTTCue::VTTCue):
(WebCore::VTTCue::setVertical):
(WebCore::VTTCue::setLine):
(WebCore::VTTCue::setLineAlign):
(WebCore::VTTCue::position const):
(WebCore::VTTCue::setPositionAlign):
(WebCore::VTTCue::setAlign):
(WebCore::VTTCue::calculateComputedLinePosition const):
(WebCore::VTTCue::calculateComputedTextPosition const):
(WebCore::VTTCue::calculateComputedPositionAlignment const):
(WebCore::VTTCue::calculateMaximumSize const):
(WebCore::VTTCue::calculateDisplayParameters):
(WebCore::VTTCue::obtainCSSBoxes):
(WebCore::VTTCue::updateDisplayTree):
(WebCore::VTTCue::getDisplayTree):
(WebCore::VTTCue::getPositionCoordinates const):
(WebCore::VTTCue::setCueSettings):
(WebCore::VTTCue::getCSSAlignment const):
(WebCore::VTTCue::getCSSWritingMode const):
(WebCore::VTTCue::setFontSize):
(WebCore::VTTCue::toJSON const):
(WebCore::horizontalKeyword): Deleted.
(WebCore::VTTCue::initialize): Deleted.
(WebCore::VTTCue::vertical const): Deleted.
(WebCore::VTTCue::lineAlign const): Deleted.
(WebCore::VTTCue::positionAlign const): Deleted.
(WebCore::VTTCue::align const): Deleted.
* Source/WebCore/html/track/VTTCue.h:
(WebCore::VTTCue::vertical const):
(WebCore::VTTCue::lineAlign const):
(WebCore::VTTCue::positionAlign const):
(WebCore::VTTCue::align const):
(WebCore::VTTCue::element const):
(WebCore::VTTCue::backdrop const):
(WebCore::VTTCue::fontSize const):
(WebCore::VTTCue::fontSizeIsImportant const):
(WebCore::VTTCue::left const):
(WebCore::VTTCue::top const):
(WebCore::VTTCue::width const):
(WebCore::VTTCue::height const):
(WebCore::VTTCue::getWritingDirection const): Deleted.
(WebCore::VTTCue::getAlignment const): Deleted.
* Source/WebCore/html/track/VTTCue.idl:
* Source/WebCore/rendering/RenderVTTCue.cpp:
(WebCore::RenderVTTCue::initializeLayoutParameters):
(WebCore::RenderVTTCue::placeBoxInDefaultPosition):
(WebCore::RenderVTTCue::overlappingObjectForRect const):
(WebCore::RenderVTTCue::shouldSwitchDirection const):
(WebCore::RenderVTTCue::moveBoxesByStep):
(WebCore::RenderVTTCue::findNonOverlappingPosition const):
(WebCore::RenderVTTCue::repositionCueSnapToLinesSet):
(WebCore::RenderVTTCue::repositionGenericCue):
(WebCore::RenderVTTCue::repositionCueSnapToLinesNotSet):
* Tools/TestWebKitAPI/Tests/WTF/EnumTraits.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/265596@main
